app/session.py

```diff
@@ -0,0 +1,90 @@
+import uuid
+
+import flask
+import redis
+
+try:
+    import cPickle as pickle
+except ImportError:
+    import pickle
+
+import itsdangerous
+from flask.sessions import SessionMixin, SessionInterface
+from werkzeug.datastructures import CallbackDict
+
+SESSION_PREFIX = "session"
+
+
+class ServerSession(CallbackDict, SessionMixin):
+    def __init__(self, initial=None, sid=None):
+        def on_update(self):
+            self.modified = True
+
+        CallbackDict.__init__(self, initial, on_update)
+        self.sid = sid
+        self.modified = False
+
+
+class RedisSessionStore(SessionInterface):
+    def __init__(self, redis, app):
+        self._redis = redis
+        self._app = app
+
+    def _get_signer(self, app) -> itsdangerous.Signer:
+        return itsdangerous.Signer(
+            app.secret_key, salt="session", key_derivation="hmac"
+        )
+
+    def _get_key(self, sid: str) -> str:
+        return f"{SESSION_PREFIX}:{sid}"
+
+    def open_session(self, app: flask.Flask, request: flask.Request):
+        unverified_sid = request.cookies.get(app.session_cookie_name)
+        if not unverified_sid:
+            unverified_sid = str(uuid.uuid4())
+            return ServerSession(sid=unverified_sid)
+        signer = self._get_signer(app)
+        try:
+            sid_as_bytes = signer.unsign(unverified_sid)
+            sid = sid_as_bytes.decode()
+        except itsdangerous.BadSignature:
+            unverified_sid = str(uuid.uuid4())
+            return ServerSession(sid=unverified_sid)
+
+        val = self._redis.get(self._get_key(sid))
+        if val is not None:
+            try:
+                data = pickle.loads(val)
+                return ServerSession(data, sid=sid)
+            except:
+                pass
+        return ServerSession(sid=unverified_sid)
+
+    def save_session(
```

Comment:
do you know where is the part a session is removed when user logs out?

Reply:
Based on this implementation, I'd say that if the session is not in Redis, a `ServerSession` is returned anyways, but without the `data` parameter.

Reply:
I've added an explicit deletion of the session data from redis just to be on the safe side